src/coreclr/tools/dotnet-pgo/TypeRefTypeSystem/TypeRefTypeSystemModule.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using Internal.TypeSystem;
+
+namespace Microsoft.Diagnostics.Tools.Pgo.TypeRefTypeSystem
+{
+    class TypeRefTypeSystemModule : ModuleDesc, IAssemblyDesc
+    {
+        AssemblyName _name;
+        List<TypeRefTypeSystemType> _types = new List<TypeRefTypeSystemType>();
+        Dictionary<string, TypeRefTypeSystemType> _nonNamespacedTypes = new Dictionary<string, TypeRefTypeSystemType>();
+        Dictionary<string, Dictionary<string, TypeRefTypeSystemType>> _namespacedTypes = new Dictionary<string, Dictionary<string, TypeRefTypeSystemType>>();
+
+        public TypeRefTypeSystemModule(TypeSystemContext tsc, AssemblyName name) : base(tsc, null)
+        {
+            _name = name;
+        }
+
+        public TypeRefTypeSystemType GetOrAddType(string nameSpace, string name)
+        {
+            TypeRefTypeSystemType type = GetTypeInternal(nameSpace, name);
+            if (type == null)
+            {
+                type = new TypeRefTypeSystemType(nameSpace, name, this);
+
+                Dictionary<string, TypeRefTypeSystemType> nameToTypeDictionary = _nonNamespacedTypes;
+                if (nameSpace != null)
```

Comment:
Nit - do we have a guarantee we don't need to use <code>string.IsNullOrEmpty</code> here?

Reply:
Yes, the only user will pass null here, but I've changed it to IsNullOrEmpty as that matches the other paths around this logic.

Reply:
Yeah, null vs. empty string has always been a nasty dilemma I kept hitting years before Microsoft when I used to make my living by developing custom database client apps ;-).